controllers/class.tapestry-controller.php

```diff
@@ -0,0 +1,337 @@
+        }
+        if (!empty($tapestry->links)) {
+            return TapestryErrors::throwsError('LINKS_EXIST_IN_NEW_TAPESTRY');
+        }
+
+        return $this->_addTapestry($tapestry);
+    }
+
+    /**
+     * Retrieve a Tapestry post
+     * 
+     * @return  Object  $tapestry
+     */
+    public function get()
+    {
+        if (!$this->postId) {
+            return TapestryErrors::throwsError('INVALID_POST_ID');
+        }
+
+        return $this->_getTapestry();
+    }
+
+    /**
+     * Add a new node
+     * 
+     * @param   Object  $node   Tapestry node
+     * 
+     * @return  Object  $node   Tapestry node
+     */
+    public function addNode($node)
+    {
+        $tapestryNodeController = new TapestryNodeController($this->postId);
+        $node = $tapestryNodeController->save($node);
+
+        if (is_wp_error($node)) {
+            return $node;
+        }
+
+        $tapestry = get_post_meta($this->postId, 'tapestry', true);
+
+        if (!isset($tapestry->nodes)) {
+            $tapestry->nodes = [];
+        }
+
+        array_push($tapestry->nodes, $node->id);
+
+        if (empty($tapestry->rootId)) {
+            $tapestry->rootId = $tapestry->nodes[0];
+        }
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
+
+        return $node;
+    }
+
+    /**
+     * Add a new link
+     * 
+     * @param  Object   $link   Tapestry link
+     * 
+     * @return  Object  $link   Tapestry link
+     */
+    public function addLink($link)
+    {
+        $tapestryLinkController = new TapestryLinkController($this->postId);
+        $link = $tapestryLinkController->save($link);
+
+        return $link;
+    }
+
+    /**
+     * Add a new group
+     * 
+     * @param   Object  $group   Tapestry group
+     * 
+     * @return  Object  $group   Tapestry group
+     */
+    public function addGroup($group)
+    {
+        $tapestryGroupController = new TapestryGroupController($this->postId);
+        $group = $tapestryGroupController->save($group);
+
+        if (is_wp_error($group)) {
+            return $group;
+        }
+
+        $tapestry = get_post_meta($this->postId, 'tapestry', true);
+
+        if (!isset($tapestry->groups)) {
+            $tapestry->groups = [];
+        }
+
+        array_push($tapestry->groups, $group->id);
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
+
+        return $group;
+    }
+
+    /**
+     * Get the node controller with associated node meta ID
+     * 
+     * @param   Number  $nodeMetaId node meta ID
+     *
+     * @return  Object  $node       node controller
+     */
+    public function getNode($nodeMetaId)
+    {
+        return new TapestryNodeController($this->postId, $nodeMetaId);
+    }
+
+    /**
+     * Get the group controller
+     * 
+     * @return  Object  $group  group controller
+     */
+    public function getGroups()
+    {
+        return new TapestryGroupController($this->postId);
```

Comment:
An instance of TapestryGroup class should be a representation of 1 group. So if you want to get groups for this tapestry, please iterate over the groups as per `$this->groups` and load them one by one into an array that you will then return.